File: atcoder/ABC/contest_420/e/old/main.cpp
#include <iostream>
#include <vector>

class UnionFind {
private:
    std::vector<int> par;
    std::vector<int> siz;
    std::vector<int> blacks;
    std::vector<bool> colors;

public:
    UnionFind(int n): par(n+1, -1), siz(n+1, 1), blacks(n+1, 0), colors(n+1, false) {}

    int root(int v) {
        while(this->par[v] != -1)
            v = this->par[v];
        return v;
    }

    void unite(int u, int v) {
        int uRoot = this->root(u);
        int vRoot = this->root(v);

        if (uRoot == vRoot) return;
        if (uRoot > vRoot) {
            this->par[vRoot] = uRoot;
            this->siz[uRoot] += vRoot;
            this->blacks[uRoot] += this->blacks[vRoot];
        } else {
            this->par[uRoot] = vRoot;
            this->siz[vRoot] += uRoot;
            this->blacks[vRoot] += this->blacks[uRoot];
        }
    }

    void add(int v) {
        int val = 1;
        if (this->colors[v] == true)
            val = -1;
        this->colors[v] = !this->colors[v];
        this->blacks[this->root(v)] += val;
    }

    bool hasBlack(int v) {
        return this->blacks[this->root(v)] > 0;
    }

    void ft_print() {
        std::cout << "par: ";
        for (int v : par) {
            std::cout << v << " ";
        }
        std::cout << std::endl;
    }
};

int main() {
    int n, q;

    std::cin >> n >> q;
    UnionFind uf = UnionFind(n);
    for (int i = 0; i < q; i++) {
        int query;
        std::cin >> query;
        if (query == 1) {
            int u, v;
            std::cin >> u >> v;
            uf.unite(u, v);
        } else if (query == 2) {
            int u;
            std::cin >> u;
            uf.add(u);
        } else {
            int u;
            std::cin >> u;
            if (uf.hasBlack(u) == true) {
                std::cout << "Yes" << std::endl;
            } else {
                std::cout << "No" << std::endl;
            }
        }
    }

    return;
}

File: atcoder/ABC/contest_420/e/old/main.py
class UnionFind:
    def __init__(self, n):
        self.parent = [-1] * n
        self.black_count = [0] * n

    def root(self, x):
        if self.parent[x] < 0:
            return x
        self.parent[x] = self.root(self.parent[x])
        return self.parent[x]

    def unite(self, a, b):
        ra = self.root(a)
        rb = self.root(b)
        if ra == rb:
            return
        if self.parent[ra] > self.parent[rb]:
            ra, rb = rb, ra
        self.parent[ra] += self.parent[rb]
        self.parent[rb] = ra
        self.black_count[ra] += self.black_count[rb]

    def add(self, x, val):
        r = self.root(x)
        self.black_count[r] += val

    def has_black(self, x):
        r = self.root(x)
        return self.black_count[r] > 0


if __name__ == "__main__":
    n, q = map(int, input().split())
    uf = UnionFind(n)
    color = [0] * n

    ans = []
    for _ in range(q):
        s = input().split()
        t = int(s[0])
        if t == 1:
            u = int(s[1]) - 1
            v = int(s[2]) - 1
            uf.unite(u, v)
        elif t == 2:
            v = int(s[1]) - 1
            if color[v] == 0:
                color[v] = 1
                uf.add(v, 1)
            else:
                color[v] = 0
                uf.add(v, -1)
        else:
            v = int(s[1]) - 1
            if uf.has_black(v):
                ans.append("Yes")
            else:
                ans.append("No")

    for a in ans:
        print(a)
